sqlmesh/core/linter/definition.py

```
@@ -0,0 +1,72 @@
+
+    rules = set()
+    for rule_name in rule_names:
+        if rule_name not in all_rules:
+            raise_config_error(f"Rule {rule_name} could not be found")
+
+        rules.add(all_rules[rule_name])
+
+    return RuleSet(rules)
+
+
+class Linter:
+    def __init__(
+        self, enabled: bool, all_rules: RuleSet, rules: RuleSet, warn_rules: RuleSet
+    ) -> None:
+        self.enabled = enabled
+        self.all_rules = all_rules
+        self.rules = rules
+        self.warn_rules = warn_rules
+
+    @classmethod
+    def from_rules(cls, all_rules: RuleSet, config: LinterConfig) -> Linter:
+        ignored_rules = select_rules(all_rules, config.ignored_rules)
+        included_rules = all_rules.difference(ignored_rules)
+
+        rules = select_rules(included_rules, config.rules)
+        warn_rules = select_rules(included_rules, config.warn_rules)
+
+        if overlapping := rules.intersection(warn_rules):
+            overlapping_rules = ", ".join(rule for rule in overlapping)
+            raise_config_error(
+                f"Rules cannot simultaneously warn and raise an error: [{overlapping_rules}]"
+            )
+
+        return Linter(config.enabled, all_rules, rules, warn_rules)
+
+    def lint_model(self, model: Model) -> None:
+        if not self.enabled:
+            return
+
+        ignored_rules = select_rules(self.all_rules, model.ignored_rules)
+
+        rules = self.rules.difference(ignored_rules)
+        warn_rules = self.warn_rules.difference(ignored_rules)
+
+        error_violations = rules.check_model(model)
+        warn_violations = warn_rules.check_model(model)
+
+        if warn_violations:
+            warn_msg = "\n".join(f" - {warn_violation}" for warn_violation in warn_violations)
+            get_console().log_warning(f"Linter warnings for {model._path}:\n{warn_msg}")
+
+        if error_violations:
+            error_msg = "\n".join(f" - {error_violations}" for error_violations in error_violations)
```

The error formatting should be handled in the `console` object and not here. Instead we should raise an exception that contains violations and the model path / name and then let console render it correctly.

---

To make sure we're aligned: Afaict we should move the logic to an abstract method in `console.py::Console` and implement it in `console.py::TerminalConsole`, simply moving this formatting there.

However, I'm not sure if I follow the exception part, it looks to me like you had something like `Console::log_failed_models()` call in `evaluator.py` in mind but the Linter rules do not raise at their current state.

From what I can tell, a happy medium (albeit with an awkward `ConfigError`) would look like:

```
if warn_violations:
  get_console().show_linter_violations(warn_violations, model.path, error=False)

if error_violations:
   get_console().show_linter_violations(error_violations, model.path, error=True)
   raise ConfigError() # empty?
```

---

Yes, the interface is correct. Exception can be:
```
raise ConfigError("Linter detected errors in the code. Please fix them before proceeding.")
```

Additionally, we should print out ALL warnings / errors for ALL models before raising.